submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,136 @@
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    // ClusterSelector identifies the Cluster resources representing a subset of clusters. not used in combination with
+    // LeftClusterSelector or RightClusterSelector
+    // An empty selector indicates wildcard, ie matches any cluster.
+    ClusterSelector metav1.LabelSelector `json:"clusterSelector,omitempty"`
+
+    CableDriver *CableDriverSpec `json:"cableDriver,omitempty"`
+    
+    priority  *int     `json:"priority,omitempty"`
+}
+
+type CableDriverSpec struct {
+    // Name of the cable driver implementation.
+    Name string `json:"name"`
+
+    // Options specifies key/value configuration parameters for the cable driver.
+    Options map[string]string `json:"options,omitempty"`
+}
+```
+
+The `ClusterConnectionPolicies` CR defines the criteria for selecting which cluster pairs/subsets can connect and associated optional
+configuration options to use for each cluster connection. To resolve the connectivity policy for a remote cluster's Endpoint,
+Submariner would search for a policy whose left and right label selectors match the corresponding Cluster resources of the local
+and remote Endpoints. If any matching policy is found, the clusters are connected. If more than one matching policy is found, the
+priority field is used to determine which policy to use. If no matching policy is found, the default cable driver and policy (full mesh)
+are used.
+
+Defining separate resources for each policy and defining a priority or precedence field, allows Submariner to prioritize the
+connectivity and cable type between a pair of clusters.
+
+The `ClusterConnectionPolicies` resources would be created and maintained on the broker cluster and synced to each participating cluster.
+
+#### Examples
+
+- We want to use a specific cable driver to connect on-premise clusters. We define a policy to select clusters labeled as
+`on-premise` to use a `vxlan` cable driver. The default cable driver will be used to connect everything else.
+
+```yaml
+kind: ClusterConnectionPolicies
+ metadata:
+   name: aws-security-zone
+ spec:
+   leftClusterSelector:
+     location: on-premise
+   rightClusterSelector:
+     location: on-premise
+   cableDriver:
+      name:  "vxlan"
+   priority: 3
+```
+
+Alternatively the example above could use the clusterSelector field as both clusters will have the same label.
+
+```yaml
+kind: ClusterConnectionPolicies
+ metadata:
+   name: aws-security-zone
+ spec:
+   clusterSelector:
+     location: on-premise
+   cableDriver:
+      name:  "vxlan"
+   priority: 3
+```
+
+- We want to sepecify the default cable driver for the full mesh policy.
+
+```yaml
+kind: ClusterConnectionPolicies
+ metadata:
+   name: public-internet
+ spec:
+   cableDriver: 
+      name: "gre+ipsec"
+   priority: 0
+```
```

Comment:
We need to propose a `subctl` command-line interface and define where policies can be created. My first thought is that we will need:

- a way to define and manage policy e.g. `subctl policy add | list | delete`
- a way to label clusters, maybe as part of join, maybe by naming the desired policy?
- other convenience shortcuts?

It may be sufficient to say that the policy is mastered on the broker and is synced to clusters. Need to think about whether any use cases also need a cluster admin to be able to create policy and have it synced to broker.

Reply:
Yeah - I will add something for this

Reply:
I've added a little more detail on the subctl commands... The last bit to go is the cable driver config... I will do something for that this afternoon